File: source/python/parallel_merge_sort.py
import threading

def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

def merge_sort(arr, result_holder):
    if len(arr) <= 1:
        result_holder.append(arr)
        return

    mid = len(arr) // 2
    left_result = []
    right_result = []

    left_thread = threading.Thread(target=merge_sort, args=(arr[:mid], left_result))
    right_thread = threading.Thread(target=merge_sort, args=(arr[mid:], right_result))

    left_thread.start()
    right_thread.start()
    left_thread.join()
    right_thread.join()

    result_holder.append(merge(left_result[0], right_result[0]))

arr = [38, 27, 43, 3, 9, 82, 10, 55]
result = []
merge_sort(arr, result)
print(f"sorted: {' '.join(str(x) for x in result[0])}")

File: source/rust/parallel_merge_sort.rs
use std::thread;

fn merge(left: Vec<i32>, right: Vec<i32>) -> Vec<i32> {
    let mut result = Vec::new();
    let mut i = 0; let mut j = 0;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] { result.push(left[i]); i += 1; }
        else { result.push(right[j]); j += 1; }
    }
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);
    result
}

fn merge_sort(arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 { return arr; }
    let mid = arr.len() / 2;
    let left_arr = arr[..mid].to_vec();
    let right_arr = arr[mid..].to_vec();
    let left_thread = thread::spawn(move || merge_sort(left_arr));
    let right_thread = thread::spawn(move || merge_sort(right_arr));
    let sorted_left = left_thread.join().unwrap();
    let sorted_right = right_thread.join().unwrap();
    merge(sorted_left, sorted_right)
}

fn main() {
    let arr = vec![38, 27, 43, 3, 9, 82, 10, 55];
    let sorted = merge_sort(arr);
    let s: Vec<String> = sorted.iter().map(|x| x.to_string()).collect();
    println!("sorted: {}", s.join(" "));
}
